### PR TITLE
Igus vision and cam space vision

### PR DESCRIPTION
**Pull request by JakeFountain**

### CODE REVIEW

File: module/vision/GoalDetector/src/GoalDetector.cpp [hunk 1]
```diff
@@ -248,93 +294,191 @@ namespace vision {
...
-                        // Check that our two goal lines are approximatly parallel
-                        && std::abs(arma::dot(lhs, rhs)) > MAXIMUM_ANGLE_BETWEEN_GOALS;
+                    bool closeToVisualHorizon = utility::vision::visualHorizonAtPoint(image, it->quad.bl[0])
+                                                    < it->quad.bl[1] + VISUAL_HORIZON_BUFFER
+                                                || utility::vision::visualHorizonAtPoint(image, it->quad.br[0])
+                                                       < it->quad.br[1] + VISUAL_HORIZON_BUFFER;
 
-                    // Check that our goals don't form too much of an upward cup (not really possible for us)
-                    //&& lhs.at(0) * rhs.at(1) - lhs.at(1) * rhs.at(0) > MAXIMUM_VERTICAL_GOAL_PERSPECTIVE_ANGLE;
+                    // TODO NEEDED?: Check we finish above the kinematics horizon or kinematics
```
Comment (TrentHouliston) [inline module/vision/GoalDetector/src/GoalDetector.cpp]: Make sure you check this todo before merging

File: module/vision/GoalDetector/src/GoalDetector.cpp [hunk 2]
```diff
@@ -248,93 +294,191 @@ namespace vision {
...
+                    // horizon is off the screen
 
+                    bool aboveHorizon = (arma::dot(convert<double, 3>(image.horizon_normal), ctr) > 0)
+                                        || (arma::dot(convert<double, 3>(image.horizon_normal), ctl) > 0);
+
+                    // Check that our two goal lines are perpendicular with the horizon must use greater
+                    // than rather then less than because of the cos
+
```
Comment (TrentHouliston) [inline module/vision/GoalDetector/src/GoalDetector.cpp]: blank line between comment and code is confusing

File: module/vision/GoalDetector/src/GoalDetector.cpp [hunk 3]
```diff
@@ -398,11 +542,11 @@ namespace vision {
...
-                    arma::vec3 cbl       = getCamFromScreen(bl, cam.focalLengthPixels);
-                    arma::vec3 cbr       = getCamFromScreen(br, cam.focalLengthPixels);
+                    // Get vectors for TL TR BL BR;
+                    arma::vec3 ctl       = convert<double, 3>(it->frustum.tl);
+                    arma::vec3 ctr       = convert<double, 3>(it->frustum.tr);
+                    arma::vec3 cbl       = convert<double, 3>(it->frustum.bl);
+                    arma::vec3 cbr       = convert<double, 3>(it->frustum.br);
                     arma::vec3 rGCc_norm = arma::normalise((cbl + cbr) * 0.5);  // vector to bottom centre of goal post
 
                     // TODO: NORMALS not used currently - delete?
```
Comment (TrentHouliston) [inline module/vision/GoalDetector/src/GoalDetector.cpp]: Delete this todo, the corners are good enough to calculate the normals if needed

File: module/vision/GoalDetector/src/GoalDetector.cpp [hunk 4]
```diff
@@ -469,9 +613,9 @@ namespace vision {
                         it->measurement.push_back(Goal::Measurement(Goal::MeasurementType::TOP_NORMAL, top));
                     }
 
+                    // TODO FIX for Igus Angular positions from the camera
```
Comment (TrentHouliston) [inline module/vision/GoalDetector/src/GoalDetector.cpp]: Is this todo important?

File: module/vision/GoalDetector/src/RansacGoalModel.h [hunk 5]
```diff
@@ -24,69 +24,59 @@
...
+        RansacGoalModel() : leftPlane(), rightPlane() {}
 
         bool regenerate(const std::array<DataPoint, REQUIRED_POINTS>& pts);
 
         double calculateError(const DataPoint& p) const;
 
         template <typename Iterator>
         void refineModel(Iterator& begin, Iterator& end, const double& threshold) {
-
+            // TODO: new least squares model for plane
```
Comment (TrentHouliston) [inline module/vision/GoalDetector/src/RansacGoalModel.h]: This made a big difference in the old vision system to the quality of goal measurements.
Might be another PR though, make an issue for it if you don't do it now.

File: module/vision/IgusVisionTests/src/IgusVisionTests.cpp [hunk 6]
```diff
@@ -0,0 +1,188 @@
+#include "IgusVisionTests.h"
```
Comment (TrentHouliston) [inline module/vision/IgusVisionTests/src/IgusVisionTests.cpp]: This should be a unit test!

File: module/vision/LUTClassifier/data/config/LUTClassifier.yaml [hunk 7]
```diff
@@ -17,14 +17,14 @@ ball:
...
-  # When clustering goals on the X axis, if a gap of at least this amount is detected it will be a new cluster
-  maximum_vertical_cluster_spacing: 0.1
   # A multiplier for spacing between ball finder lines (based on the closeness)
   horizontal_subsample_factor: 0.5
   # The longest pixel count we can strike a lighitng bolt around things
   maximum_lighting_bolt_length: 20
   # The minimum magnitude of gradient a lighting bolt can have before it stops
   minimum_lighting_bolt_strength: 15.0
+  # TODO: Definition
+  maximum_vertical_cluster_spacing: 10
```
Comment (TrentHouliston) [inline module/vision/LUTClassifier/data/config/LUTClassifier.yaml]: Just delete this, it isn't used

File: module/vision/LUTClassifier/src/GoalEnhancer.cpp [hunk 8]
```diff
@@ -87,6 +89,7 @@ namespace vision {
          */
 
         Line horizon(convert<double, 2>(classifiedImage.horizon.normal), classifiedImage.horizon.distance);
+        // Plane<3> horizon;
```
Comment (TrentHouliston) [inline module/vision/LUTClassifier/src/GoalEnhancer.cpp]: Should use the real horizon for this mini ransac through centre points, otherwise it'll end up classifying a whole lot of junk with very dense classification lines.

File: module/vision/LUTClassifier/src/VisualHorizonFinder.cpp [hunk 9]
```diff
@@ -31,24 +33,34 @@ namespace vision {
...
 
     void LUTClassifier::findVisualHorizon(const Image& image,
                                           const LookUpTable& lut,
-                                          ClassifiedImage& classifiedImage) {
+                                          ClassifiedImage& classifiedImage,
+                                          const CameraParameters& cam) {
 
         // Get some local references to class variables to make text shorter
-        Line horizon(convert<double, 2>(classifiedImage.horizon.normal), classifiedImage.horizon.distance);
+        // Line horizon(convert<double, 2>(classifiedImage.horizon.normal), classifiedImage.horizon.distance);
```
Comment (TrentHouliston) [inline module/vision/LUTClassifier/src/VisualHorizonFinder.cpp]: Delete if unused

File: shared/message/Geometry.proto [hunk 10]
```diff
@@ -49,3 +49,15 @@ message Quad {
...
+message Frustum {
+    vec3 tl = 1;
+    vec3 tr = 2;
+    vec3 bl = 3;
+    vec3 br = 4;
+}
+
+message Cone {
+    vec3 axis = 1;
+    float gradient = 2;
```
Comment (TrentHouliston) [inline shared/message/Geometry.proto]: Everything else in this file is a double but this is a float! (probably doesn't matter)

File: shared/utility/math/geometry/Cone.h [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+            }
+
+            float distanceToPoint(const Vector& p) const {
+                Vector p_cone = projectPoint(p);
+                Vector p_orth = p - p_cone;
+                // Compute size of orth component
+                return arma::norm(p_orth);
+            }
+
+            float dotDistanceToPoint(const Vector& p) const {
```
Comment (TrentHouliston) [inline shared/utility/math/geometry/Cone.h]: Put a comment on this function as it's confusing what it does

File: shared/utility/math/ransac/RansacConeModel.h [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            static constexpr size_t REQUIRED_POINTS = 3;
+            using DataPoint                         = arma::vec3;
+
+            bool regenerate(const std::array<DataPoint, REQUIRED_POINTS>& points);
+
+            double calculateError(const DataPoint& p) const;
+
+            template <typename Iterator>
+            void refineModel(Iterator&, Iterator&, const double&) {
+                // commented out due to performance concerns - works well though
```
Comment (TrentHouliston) [inline shared/utility/math/ransac/RansacConeModel.h]: Might be worth putting this in again since it's not too hard to calculate

File: module/vision/BallDetector/src/BallDetector.cpp [hunk 13]
```diff
@@ -199,103 +211,120 @@ namespace vision {
...
                         // Transform our centre into kinematics coordinates
-                        arma::vec2 centre = imageToScreen(result.model.centre, convert<uint, 2>(image.dimensions));
+                        arma::vec3 axis = result.model.unit_axis;
 
                         // Get the 4 points around our circle
-                        arma::vec2 top   = centre + arma::vec2({0, result.model.radius});
-                        arma::vec2 base  = centre + arma::vec2({0, -result.model.radius});
-                        arma::vec2 left  = centre + arma::vec2({result.model.radius, 0});
-                        arma::vec2 right = centre + arma::vec2({-result.model.radius, 0});
+                        arma::vec2 top   = projectCamSpaceToScreen(result.model.getTopVector(), cam);
```
Comment (JakeFountain) [inline module/vision/BallDetector/src/BallDetector.cpp]: they are used and important for head behaviour

File: module/vision/IgusVisionTests/src/IgusVisionTests.cpp (hunk 6 above)
Comment (JakeFountain) [inline module/vision/IgusVisionTests/src/IgusVisionTests.cpp]: Not flexible enough. I think.

File: shared/utility/math/ransac/RansacConeModel.h (hunk 12 above)
Comment (JakeFountain) [inline shared/utility/math/ransac/RansacConeModel.h]: Another PR
